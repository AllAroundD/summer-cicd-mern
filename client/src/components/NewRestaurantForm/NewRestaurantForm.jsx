import React, { useState } from "react";
import axios from "axios";
import validatePhoneNumber from "../../utils/validatePhoneNumber";

const NewRestaurantForm = ({ getRestaurants }) => {
  const [newRestaurant, setNewRestaurant] = useState({
    name: "",
    address: "",
    phone: "",
    cuisine: "",
  });
  const [error, setError] = useState("");
  const [msg, setMessage] = useState("");

  const status = document.querySelector(".status");

  const isInvalid =
    newRestaurant.name === "" ||
    newRestaurant.address === "" ||
    newRestaurant.phone === "" ||
    !validatePhoneNumber(newRestaurant.phone) ||
    newRestaurant.cuisine === "";

  const onChange = (name) => {
    return ({ target: { value } }) => {
      setNewRestaurant((prevRestaurant) => ({
        ...prevRestaurant,
        [name]: value,
      }));
    };
  };

  const saveFormData = async () => {
    axios
      .post("/api/restaurants", newRestaurant)
      .then((response) => {
        // console.log(response.data);
        // setMessage("Your restaurant was successfully submitted!");
        status.innerText = "Your restaurant was successfully submitted!";
        // alert("Your restaurant was successfully submitted!");
        setNewRestaurant({ name: "", address: "", phone: "", cuisine: "" });
        setError("");
        getRestaurants();
        document.getElementById("name").focus();
      })
      .catch((err) => {
        console.log(err);
        setError(err.message);
      });
  };

  const onSubmit = async (e) => {
    e.preventDefault();

    try {
      const response = await saveFormData();
      // window.location.reload();
    } catch (err) {
      // alert(`Restaurant submission failed! ${err.message}`);
      setError(`Restaurant submission failed! ${err.message}`);
    }
  };

  return (
    <form onSubmit={onSubmit}>
      <div>
        {error && (
          <p data-testid="error" style={{ color: "red", fontWeight: "bold" }}>
            {error}
          </p>
        )}
        <p
          className="status"
          data-testid="status"
          style={{ color: "green", fontWeight: "bold" }}
        >
          {/* {msg} */}
        </p>
      </div>
      <div className="mb-3">
        <label htmlFor="name" className="form-label">
          Name: *
        </label>
        <input
          type="text"
          className="form-control"
          id="name"
          data-testid="restaurant-name"
          name="name"
          value={newRestaurant.name}
          onChange={onChange("name")}
          required
        />
        <label htmlFor="address" className="form-label">
          Address: *
        </label>
        <input
          type="text"
          className="form-control"
          id="address"
          data-testid="restaurant-address"
          name="address"
          value={newRestaurant.address}
          onChange={onChange("address")}
          required
        />
        <label htmlFor="phone" className="form-label">
          Phone: *
        </label>
        <input
          type="text"
          className="form-control"
          id="phone"
          data-testid="restaurant-phone"
          name="phone"
          value={newRestaurant.phone}
          maxLength="12"
          onChange={onChange("phone")}
          required
        />
        <label htmlFor="cuisine" className="form-label">
          Cuisine: *
        </label>
        <input
          type="text"
          className="form-control"
          id="cuisine"
          data-testid="restaurant-cuisine"
          name="cuisine"
          value={newRestaurant.cuisine}
          onChange={onChange("cuisine")}
          required
        />
      </div>
      <div className="text-center">
        <button
          type="submit"
          className="btn btn-primary"
          disabled={isInvalid}
          data-testid="submit"
        >
          Submit
        </button>
      </div>
    </form>
  );
};

export default NewRestaurantForm;
