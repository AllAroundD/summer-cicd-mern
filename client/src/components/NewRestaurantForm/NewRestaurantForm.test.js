import React from "react";
import { render, cleanup, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import NewRestaurantForm from "./NewRestaurantForm";

afterEach(cleanup);

describe("NewRestaurantForm", () => {
  it("updates the input value on change", () => {
    const { getByTestId } = render(<NewRestaurantForm />);
    const nameInput = getByTestId("restaurant-name");
    const addressInput = getByTestId("restaurant-address");
    const phoneInput = getByTestId("restaurant-phone");
    const cuisineInput = getByTestId("restaurant-cuisine");
    fireEvent.change(nameInput, { target: { value: "Waffle House" } });
    expect(nameInput.value).toBe("Waffle House");
    fireEvent.change(addressInput, { target: { value: "1212 Yonge Street" } });
    expect(addressInput.value).toBe("1212 Yonge Street");
    fireEvent.change(phoneInput, { target: { value: "888-888-8889" } });
    expect(phoneInput.value).toBe("888-888-8889");
    fireEvent.change(cuisineInput, { target: { value: "Chinese" } });
    expect(cuisineInput.value).toBe("Chinese");
    fireEvent.click(getByTestId("submit"));
  });

  it("matches snapshot", () => {
    const { asFragment } = render(<NewRestaurantForm />);
    expect(asFragment()).toMatchSnapshot();
  });
});
