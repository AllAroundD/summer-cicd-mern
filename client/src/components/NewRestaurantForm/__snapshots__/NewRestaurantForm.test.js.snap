// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewRestaurantForm matches snapshot 1`] = `
<DocumentFragment>
  <form>
    <div>
      <p
        class="status"
        data-testid="status"
        style="color: green; font-weight: bold;"
      />
    </div>
    <div
      class="mb-3"
    >
      <label
        class="form-label"
        for="name"
      >
        Name: *
      </label>
      <input
        class="form-control"
        data-testid="restaurant-name"
        id="name"
        name="name"
        required=""
        type="text"
        value=""
      />
      <label
        class="form-label"
        for="address"
      >
        Address: *
      </label>
      <input
        class="form-control"
        data-testid="restaurant-address"
        id="address"
        name="address"
        required=""
        type="text"
        value=""
      />
      <label
        class="form-label"
        for="phone"
      >
        Phone: *
      </label>
      <input
        class="form-control"
        data-testid="restaurant-phone"
        id="phone"
        maxlength="12"
        name="phone"
        required=""
        type="text"
        value=""
      />
      <label
        class="form-label"
        for="cuisine"
      >
        Cuisine: *
      </label>
      <input
        class="form-control"
        data-testid="restaurant-cuisine"
        id="cuisine"
        name="cuisine"
        required=""
        type="text"
        value=""
      />
    </div>
    <div
      class="text-center"
    >
      <button
        class="btn btn-primary"
        data-testid="submit"
        disabled=""
        type="submit"
      >
        Submit
      </button>
    </div>
  </form>
</DocumentFragment>
`;
